core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java

```diff
@@ -0,0 +1,174 @@
+import io.questdb.cairo.TableReader;
+import io.questdb.cairo.sql.*;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.std.Misc;
+
+public class DistinctSymbolRecordCursorFactory implements RecordCursorFactory {
+    private final DistinctSymbolRecordCursor cursor;
+    private final CairoEngine engine;
+    private final GenericRecordMetadata metadata;
+    private final CharSequence tableName;
+    private final int columnIndex;
+    private final long tableVersion;
+
+    public DistinctSymbolRecordCursorFactory(
+            final CairoEngine engine,
+            final GenericRecordMetadata metadata,
+            final CharSequence tableName,
+            final int columnIndex,
+            final long tableVersion) {
+        this.engine = engine;
+        this.metadata = metadata;
+        this.tableName = tableName;
+        this.columnIndex = columnIndex;
+        this.tableVersion = tableVersion;
+        this.cursor = new DistinctSymbolRecordCursor();
+    }
+
+    @Override
+    public void close() {
+        cursor.close();
+    }
+
+    @Override
+    public RecordCursor getCursor(SqlExecutionContext executionContext) {
+        TableReader reader = engine.getReader(executionContext.getCairoSecurityContext(), tableName, tableVersion);
+        cursor.of(reader, columnIndex);
+        return cursor;
+    }
+
+    @Override
+    public RecordMetadata getMetadata() {
+        return metadata;
+    }
+
+    @Override
+    public boolean isRandomAccessCursor() {
+        return true;
+    }
+
+    private static class DistinctSymbolRecordCursor implements RecordCursor {
+        private DistinctSymbolRecord recordA = new DistinctSymbolRecord();
+        private DistinctSymbolRecord recordB = null;
+        private TableReader reader;
+        private int columnIndex;
+        private int numberOfSymbols;
+        private SymbolMapReader symbolMapReader;
+
+        @Override
+        public void close() {
+            reader = Misc.free(reader);
+        }
+
+        @Override
+        public Record getRecord() {
+            return recordA;
+        }
+
+        @Override
+        public SymbolTable getSymbolTable(int columnIndex) {
+            return symbolMapReader;
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (recordA.getRecordIndex() + 1 < numberOfSymbols) {
+                recordA.incrementRecordIndex();
+                return true;
+            }
+            return false;
+        }
+
+        @Override
+        public Record getRecordB() {
+            if (recordB == null) {
+                recordB = new DistinctSymbolRecord();
+            }
+            recordB.reset();
+            return recordB;
+        }
+
+        @Override
+        public void recordAt(Record record, long atRowId) {
+            ((DistinctSymbolRecord) record).recordIndex = (int) atRowId;
+        }
+
+        @Override
+        public void toTop() {
+            recordA.recordIndex = -1;
```

Comment:
minor: toTop() and record.reset() implement the same logic, perhaps `topTop()` can call `recordA.reset()`. Or may be `record.reset()` could be renamed to `record.toTop()`